krun/mail.py

```diff
@@ -0,0 +1,86 @@
+from collections import deque
+from email.mime.text import MIMEText
+from datetime import timedelta, datetime
+import smtplib
+import socket
+import textwrap
+
+
+class Mailer(object):
+
+    FROM_USER = "noreply"
+    SMTP_HOST = "localhost"
+    WRAP_THRESHOLD = 72
+    RULE = WRAP_THRESHOLD * "#"
+
+    QUOTA_THRESHOLD_TEMPLATE = \
+        "krun is configured to send no more than %d mails in any given " + \
+        "one hour interval. krun is currently hard up against this limit " + \
+        "and thus may not send emails for subsequent errors. Please check " + \
+        "the log file on the benchmark system for further errors. Thanks"
+
+    def __init__(self, recipients=[], max_mails_per_hour=5):
+        self.recipients = recipients
+        self.fqdn = socket.getfqdn()
+        self.max_mails_per_hour = max_mails_per_hour
+
+        # A collection of timestamps (as timedelta instances) for mails
+        # sent in the last hour
+        self.last_mail_times = deque()
+
+    def set_recipients(self, recipients):
+        self.recipients = recipients
+
+    def _wrap_para(self, txt):
+        return "\n".join(textwrap.wrap(txt, Mailer.WRAP_THRESHOLD))
+
+    def _one_hour_ago(self):
+        return datetime.now() - timedelta(hours=1)
+
+    def _clear_old(self):
+        one_hour_ago = self._one_hour_ago()
+        while True:
+            try:
+                t = self.last_mail_times.popleft()
+            except IndexError:
+                break  # nothing in deque
+            if t > one_hour_ago:
+                self.last_mail_times.appendleft(t)  # put back
```

Comment:
If I understand this correctly (and I might not...), this is still going to send *all* mails, but just throttle them to 5 an hour. If so, I think this is undesireable -- I could still wake up to 50 mails. I think it should say something like "after 5 mails in an hour, I will throw away all further mails". After all, after 5 errors, one probably has the idea that the experiment is screwed... The complete number of errors is still in the log on the server right?